src/openrct2/world/Duck.cpp

```diff
@@ -0,0 +1,398 @@
+    {
+        state = DUCK_STATE::FLY_AWAY;
+        UpdateFlyAway();
+    }
+    else
+    {
+        waterHeight <<= 4;
+        sint32 newZ = abs(z - waterHeight);
+
+        if (manhattanDistanceN <= manhattanDistance)
+        {
+            if (newZ > manhattanDistanceN)
+            {
+                newZ = z - 2;
+                if (waterHeight >= z)
+                {
+                    newZ += 4;
+                }
+                frame = 1;
+            }
+            else
+            {
+                newZ = z;
+            }
+            MoveTo(newX, newY, newZ);
+            Invalidate();
+        }
+        else
+        {
+            if (newZ > 4)
+            {
+                state = DUCK_STATE::FLY_AWAY;
+                UpdateFlyAway();
+            }
+            else
+            {
+                state = DUCK_STATE::SWIM;
+                frame = 0;
+                UpdateSwim();
+            }
+        }
+    }
+}
+
+void rct_duck::UpdateSwim()
+{
+    if (((gCurrentTicks + sprite_index) & 3) != 0) return;
+
+    uint32 randomNumber = scenario_rand();
+    if ((randomNumber & 0xFFFF) < 0x666)
+    {
+        if (randomNumber & 0x80000000)
+        {
+            state = DUCK_STATE::DOUBLE_DRINK;
+            frame = -1;
+            UpdateDoubleDrink();
+        }
+        else
+        {
+            state = DUCK_STATE::DRINK;
+            frame = -1;
+            UpdateDrink();
+        }
+    }
+    else
+    {
+        sint32 currentMonth = date_get_month(gDateMonthsElapsed);
+        if (currentMonth >= MONTH_SEPTEMBER && (randomNumber >> 16) < 218)
+        {
+            state = DUCK_STATE::FLY_AWAY;
+            UpdateFlyAway();
+        }
+        else
+        {
+            Invalidate();
+            sint32 landZ = map_element_height(x, y);
+            sint32 waterZ = (landZ >> 16) & 0xFFFF;
+            landZ &= 0xFFFF;
+
+            if (z < landZ || waterZ == 0)
+            {
+                state = DUCK_STATE::FLY_AWAY;
+                UpdateFlyAway();
+            }
+            else
+            {
+                z = waterZ;
+                randomNumber = scenario_rand();
+                if ((randomNumber & 0xFFFF) <= 0xAAA)
+                {
+                    randomNumber >>= 16;
+                    sprite_direction = randomNumber & 0x18;
+                }
+
+                sint32 direction = sprite_direction >> 3;
+                sint32 newX = x + DuckMoveOffset[direction].x;
+                sint32 newY = y + DuckMoveOffset[direction].y;
+                landZ = map_element_height(newX, newY);
+                waterZ = (landZ >> 16) & 0xFFFF;
+                landZ &= 0xFFFF;
+
+                if (z >= landZ && z == waterZ)
+                {
+                    MoveTo(newX, newY, waterZ);
+                    Invalidate();
+                }
+            }
+        }
+    }
+}
+
+void rct_duck::UpdateDrink()
+{
+    frame++;
+    if (DuckAnimationDrink[frame] == 0xFF)
```

Comment:
This is how the original assembly did it.
The difference is that 0xFF is used to indicate a change of animation where as the ones without are simple loops.
Of course it doesn't make any sense to do this when its got separate logic for each animation type.